src/compose/service-manager.coffee

```diff
@@ -0,0 +1,276 @@
+				.then(Service.fromContainer)
+
+	# Returns an array with the container(s) matching a service by appId, commit, image and environment
+	get: (service) =>
+		@getAll("io.resin.service_id=#{service.serviceId}")
+		.then (services = []) ->
+			return _.filter services, (currentService) ->
+				currentService.isSameContainer(service)
+
+	getStatus: =>
+		@getAll()
+		.then (services) =>
+			status = _.clone(@volatileState)
+			_.forEach services, (service) ->
+				status[service.containerId] ?= _.pick(service, [ 'appId', 'imageId', 'status', 'releaseId', 'commit' ])
+			return _.values(status)
+
+	getByDockerContainerId: (containerId) =>
+		@docker.getContainer(containerId).inspect()
+		.then (container) ->
+			if !container.Config.Labels['io.resin.supervised']?
+				return null
+			return Service.fromContainer(container)
+		.catchReturn(null)
+
+	waitToKill: (service, timeout) ->
+		startTime = Date.now()
+		pollInterval = 100
+		timeout = checkInt(timeout, positive: true) ? 60000
+		checkFileOrTimeout = ->
+			killme = service.killmeFullPathOnHost()
+			fs.statAsync(killme)
+			.catch (err) ->
+				throw err unless (Date.now() - startTime) > timeout
+			.then ->
+				fs.unlinkAsync(killme).catch(_.noop)
+		retryCheck = ->
+			checkFileOrTimeout()
+			.catch ->
+				Promise.delay(pollInterval).then(retryCheck)
+		retryCheck()
+
+	prepareForHandover: (service) =>
+		@get(service)
+		.then ([ svc ]) =>
+			container = @docker.getContainer(svc.containerId)
+			container.update(RestartPolicy: {})
+			.then ->
+				container.rename(name: "old_#{service.serviceName}_#{service.releaseId}")
+
+	updateReleaseId: (service, releaseId) =>
+		@get(service)
+		.then ([ svc ]) =>
+			@docker.getContainer(svc.containerId).rename(name: "#{service.serviceName}_#{releaseId}")
+
+	handover: (currentService, targetService) =>
+		# We set the running container to not restart so that in case of a poweroff
+		# it doesn't come back after boot.
+		@prepareForHandover(currentService)
+		.then =>
+			@start(targetService)
+		.then =>
+			@waitToKill(currentService, targetService.labels['io.resin.update.handover_timeout'])
+		.then =>
+			@kill(currentService)
+
+	listenToEvents: =>
+		return if @listening
+		@listening = true
+		@docker.getEvents(filters: type: [ 'container' ])
+		.then (stream) =>
+			parser = new JSONParser()
+			parser.onError = (err) ->
+				console.error('Error on docker events stream', err, err.stack)
+				stream.destroy()
+			parser.onValue = (data) =>
+				return if parser.stack.length != 0
+				if data?.status in ['die', 'start']
+					setImmediate =>
+						@getByDockerContainerId(data.id)
+						.then (service) =>
+							if service?
+								if data.status == 'die'
+									@logger.logSystemEvent(logTypes.serviceExit, { service })
+									@containerHasDied[data.id] = true
+								else if data.status == 'start' and @containerHasDied[data.id]
+									@logger.logSystemEvent(logTypes.serviceRestart, { service })
```

Comment:
Should this also clear the flag from `@containerHasDied`?

Reply:
Yes. Done.